### PR TITLE
[issue1023] create landmark base class

### PR DESCRIPTION
**Pull request by SimonDold**

### CODE REVIEW

File: src/search/landmarks/landmark_factory_h_m.cc [hunk 1]
```diff
@@ -927,8 +929,13 @@ void LandmarkFactoryHM::add_lm_node(int set_index, bool goal) {
         vector<FactPair> facts(hm_entry.fluents);
         utils::sort_unique(facts);
         assert(!facts.empty());
-        Landmark landmark(facts, false, (facts.size() > 1), goal);
-        landmark.first_achievers.insert(
+        std::shared_ptr<Landmark> landmark;
+        if ((facts.size() > 1)){
```
Comment (remochristen) [inline src/search/landmarks/landmark_factory_h_m.cc:933]: I would make the if condition `facts.size() == 1` and thus construct `SimpleLandmark`s there.

**Review by remochristen — COMMENTED** [review]
I looked through the changes and added some comments.
The overall design makes sense to me and should make it easier to add other types of landmarks down the line, disjunctive action landmarks for example.

File: src/search/landmarks/landmark.h [hunk 2]
```diff
@@ -6,17 +6,24 @@
...
+enum class LandmarkType {
+    SIMPLE = 0,
+    DISJUNCTIVE = 1,
+    CONJUNCTIVE = 2
+};
+
 class Landmark {
 public:
-    Landmark(std::vector<FactPair> _facts, bool disjunctive, bool conjunctive,
+    explicit Landmark(std::vector<FactPair> _facts,
```
[the remark this answers is not in the document]
Comment (maltehelmert) [reply src/search/landmarks/landmark.h:18]: Also, we shouldn't add `explicit` here, as the constructor has more than one argument.

File: src/search/landmarks/landmark_graph.h [hunk 3]
```diff
@@ -54,11 +54,11 @@ class LandmarkNode {
     }
 
     // TODO: Remove this function once the LM-graph is constant after creation.
-    Landmark &get_landmark() {
+    std::shared_ptr<Landmark> get_landmark() {
         return landmark;
     }
 
-    const Landmark &get_landmark() const {
+    const std::shared_ptr<Landmark> get_landmark() const {
```
[the remark this answers is not in the document]
Comment (maltehelmert) [reply src/search/landmarks/landmark_graph.h:61]: Should this be a shared pointer in the first place? Why shared ownership?

[the remark this answers is not in the document]
Comment (maltehelmert) [reply src/search/landmarks/landmark_graph.h:61]: Shared pointers fall a little bit into the "feature of last resort" category.

[the remark this answers is not in the document]
Comment (SimonDold) [reply src/search/landmarks/landmark_graph.h:61]: in `LandmarkFactoryRpgExhaust::generate_relaxed_landmarks` the landmark is given to a different class to do a check but also kept in the `LandmarkFactoryRpgExhaust` class to add it to the landmark graph if said check fails

**Comment by ClemensBuechner** [comment]
Closing this because we decided to not make this change after all.

**Pull request closed without merging**